src/boss/azure_worker.go

```diff
@@ -0,0 +1,214 @@
+package boss
+
+import (
+	"fmt"
+	"log"
+	"net/http"
+	"os"
+	"os/exec"
+	"os/user"
+	"sync"
+	"time"
+)
+
+type AzureWorkerPool struct {
+	workerNum int
+	workers   *map[string]*AzureWorker
+	nextId    int
+}
+
+type AzureWorker struct {
+	pool         *AzureWorkerPool
+	workerId     string
+	configPosit  int
+	diskName     string
+	vnetName     string
+	subnetName   string
+	nsgName      string
+	nicName      string
+	publicIPName string
+	privateAddr  string
+	publicAddr   string
+}
+
+func NewAzureWorkerPool() *WorkerPool {
+	conf, err := ReadAzureConfig()
+	if err != nil {
+		log.Fatalln(err)
+	}
+	num := conf.Resource_groups.Rgroup[0].Numvm
+	workers := make(map[string]*AzureWorker, num)
+	pool := &AzureWorkerPool{
+		workerNum: num,
+		workers:   &workers,
+		nextId:    num + 1,
+	}
+	for i := 0; i < num; i++ {
+		worker_i := new(AzureWorker)
+		worker_i.pool = pool
+		worker_i.privateAddr = *conf.Resource_groups.Rgroup[0].Vms[i].Net_ifc.Properties.IPConfigurations[0].Properties.PrivateIPAddress
+		publicWrap := conf.Resource_groups.Rgroup[0].Vms[i].Net_ifc.Properties.IPConfigurations[0].Properties.PublicIPAddress
+		if publicWrap == nil {
+			worker_i.publicAddr = ""
+		} else {
+			worker_i.publicAddr = *publicWrap.Properties.IPAddress
+		}
+		worker_i.workerId = *conf.Resource_groups.Rgroup[0].Vms[i].Vm.Name
+		worker_i.configPosit = num
+	}
+	parent := &WorkerPool{
+		WorkerPoolPlatform: pool,
+	}
+	return parent
+}
+
+// Is nextId here useful? I store nextId in the pool
+// TODO: maybe store nextId to the config file so that if the boss shut down, it know how to do next time
+func (pool *AzureWorkerPool) NewWorker(workerId string) *Worker {
+	return &Worker{
+		workerId: workerId,
+		workerIp: "",
+	}
+}
+
+// TODO: make AzureCreateVM multiple-threaded
+func (pool *AzureWorkerPool) CreateInstance(worker *Worker) error {
+	log.Printf("creating an azure worker\n")
+	conf, err := AzureCreateVM(worker)
+	if err != nil {
+		return err
+	}
+	var private string
+
+	vmNum := conf.Resource_groups.Rgroup[0].Numvm
+	private = worker.workerIp
+	newDiskName := worker.workerId + "-disk"
+	newNicName := worker.workerId + "-nic"
+	newNsgName := worker.workerId + "-nsg"
+	subnetName := worker.workerId + "-subnet"
+	vnetName := "ol-boss-vnet"
+	publicIPName := ""
+	public := ""
+
+	azworker := &AzureWorker{
+		pool:         pool,
+		workerId:     worker.workerId,
+		configPosit:  vmNum - 1,
+		diskName:     newDiskName,
+		vnetName:     vnetName,
+		nicName:      newNicName,
+		nsgName:      newNsgName,
+		subnetName:   subnetName,
+		publicIPName: publicIPName,
+		privateAddr:  private,
+		publicAddr:   public, // If newly created one, this is ""
+	}
+
+	pool.workerNum += 1
+	pool.nextId = pool.workerNum + 1
+
+	(*pool.workers)[azworker.workerId] = azworker
+	worker.workerId = azworker.workerId
+	worker.workerIp = azworker.privateAddr
+
+	return nil
+}
+
+func (worker *Worker) startWorker() {
```

Comment:
The the type is worker, then start() is a better function name than startWorker() -- the latter is a bit redundant